lightning/src/ln/channelmanager.rs

```diff
@@ -7067,6 +7078,152 @@ where
 		}
 	}
 
+	/// Creates an [`OfferBuilder`] such that the [`Offer`] it builds is recognized by the
+	/// [`ChannelManager`] when handling [`InvoiceRequest`] messages for the offer. The offer will
+	/// not have an expiration unless otherwise set on the builder.
+	///
+	/// Uses a one-hop [`BlindedPath`] for the offer with [`ChannelManager::get_our_node_id`] as the
+	/// introduction node and a derived signing pubkey for recipient privacy. As such, currently,
+	/// the node must be announced. Otherwise, there is no way to find a path to the introduction
+	/// node in order to send the [`InvoiceRequest`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+	pub fn create_offer_builder(
+		&self, description: String
+	) -> OfferBuilder<DerivedMetadata, secp256k1::All> {
+		let node_id = self.get_our_node_id();
+		let expanded_key = &self.inbound_payment_key;
+		let entropy = &*self.entropy_source;
+		let secp_ctx = &self.secp_ctx;
+		let path = self.create_one_hop_blinded_path();
+
+		OfferBuilder::deriving_signing_pubkey(description, node_id, expanded_key, entropy, secp_ctx)
+			.chain_hash(self.chain_hash)
+			.path(path)
+	}
+
+	/// Creates a [`RefundBuilder`] such that the [`Refund`] it builds is recognized by the
+	/// [`ChannelManager`] when handling [`Bolt12Invoice`] messages for the refund. The refund will
+	/// not have an expiration unless otherwise set on the builder.
+	///
+	/// The provided `payment_id` is used to ensure that only one invoice is paid for the refund.
+	///
+	/// Uses a one-hop [`BlindedPath`] for the refund with [`ChannelManager::get_our_node_id`] as
+	/// the introduction node and a derived payer id for sender privacy. As such, currently, the
+	/// node must be announced. Otherwise, there is no way to find a path to the introduction node
+	/// in order to send the [`Bolt12Invoice`].
+	///
+	/// [`Refund`]: crate::offers::refund::Refund
+	/// [`Bolt12Invoice`]: crate::offers::invoice::Bolt12Invoice
+	pub fn create_refund_builder(
+		&self, description: String, amount_msats: u64, payment_id: PaymentId, retry_strategy: Retry,
+		max_total_routing_fee_msat: Option<u64>
+	) -> Result<RefundBuilder<secp256k1::All>, Bolt12SemanticError> {
+		let node_id = self.get_our_node_id();
+		let expanded_key = &self.inbound_payment_key;
+		let entropy = &*self.entropy_source;
+		let secp_ctx = &self.secp_ctx;
+		let path = self.create_one_hop_blinded_path();
+
+		let builder = RefundBuilder::deriving_payer_id(
+			description, node_id, expanded_key, entropy, secp_ctx, amount_msats, payment_id
+		)?
+			.chain_hash(self.chain_hash)
+			.path(path);
+
+		self.pending_outbound_payments
+			.add_new_awaiting_invoice(payment_id, retry_strategy, max_total_routing_fee_msat)
```

Comment:
Is it the case that if this refund is not paid within 3 timer ticks, we won't accept invoices to it even if no expiry is set?

Reply:
Hmm... yeah, we were thinking more of the invoice request scenario where the response should be immediate. I think we'll want to add an expiration parameter and use that to determine how many timer ticks are allowed for refunds.

For the payment, we'll at least check the expiry in `OutboundPayments::find_route_and_send_payment`, but only in `std`. In general, it seems we'll always attempt to pay expired invoices in `no-std`, IIUC. 😕

Reply:
> Hm, relying on timer ticks to determine expiration sounds a bit error-prone since `timer_tick_occurred` is called "roughly 1 per minute". Maybe we should pass in the current time into `timer_tick_occurred` in `no-std` or something like that?

That's a fair point, though I'm not 100% certain how concerned we should be. We are fuzzy elsewhere with time where we are receiving money. But here we are sending money, so maybe we should be? Worse case scenario is we pay someone after user said we shouldn't. But that's only if the user wasn't calling `timer_ticker_occurred` appropriately. @TheBlueMatt Any thoughts on this?

So would we store the refund's absolute expiry in `AwaitingInvoice` and then in `std` presumably rely on `SystemTime`? For an invoice request for an offer, would we just use a hardcoded 3-minute expiration before forgetting about the request?

Note that we won't pay the invoice if _it_ is expired, not withstanding the earlier caveat about `no-std`. But it just occurred to me that my earlier statement was not exactly accurate:

> For the payment, we'll at least check the expiry in `OutboundPayments::find_route_and_send_payment`, but only in `std`. In general, it seems we'll always attempt to pay expired invoices in `no-std`, IIUC. 😕

Here, we would check the expiry of the invoice -- not the refund -- which is controlled by the refund recipient. The spec doesn't give any guidance on what to do if the refund is expired but the invoice is not. Should get this cleared up.

Reply:
I was thinking in the `Refund` case, if a merchant wants to allow a user a year to claim a refund, but they call `timer_tick_occurred` a bit too often then the user may only have like 8 months to claim in practice. But since the fix seems potentially annoying, it's probably fine to wait and see if issues actually arise?

Reply:
Yeah, the refund case is the concern. Customer requests a refund from the merchant and the merchants emails a QR code. Customer scans the QR code and sends an invoice. The time between emailing the refund QR and the customer sending the invoice could be large. It's a much different use case then when the customer scans an offer.

Reply:
Discussed offline. Updated to use `SystemTime` instead of timer ticks and block time if `no-std`. We also discussed a timer-based approach for shorter-lived refunds (i.e., less than a day) when building with `no-std`. The latest push only contains the the aforementioned approach for simplicity and to get quicker feedback. I'm not 100% clear on how we'd want to combine the timer-based approach in `AwaitingInvoice`. @TheBlueMatt Were you thinking simply having an `Option<64>` for remaining ticks? If `Some` use this in lieu of the absolute expiry?

Reply:
Hmm... ticks are essentially number of minutes left until it expires whereas the expiry is a fixed time when it expires. So they can't be compared directly unless I'm misunderstanding how they should both be represented.